lib/index.ts

```diff
@@ -0,0 +1,167 @@
+}
+
+/**
+ * Logs message with error coloring. Also includes timestamp.
+ *
+ * @param info - error to log
+ */
+function logError(error: string): void {
+    console.log('\x1b[31m%s\x1b[0m', `[${new Date().toLocaleString()}]: ${error}`);
+}
+
+/**
+ * Returns random hex color
+ */
+function generateRandomColor(): string {
+    return (
+        '#' +
+        Math.random()
+            .toString(16)
+            .slice(2, 8)
+    );
+}
+
+// Socket Connection
+
+const io = socketio();
+io.on('connection', client => {
+    /**
+     * Received when client logs in as guest and sends display name and color
+     */
+    client.on(Constants.GET_DISPLAY_NAME, () => {
+        let displayName: string = (rword.rword.generate(2, { length: '2 - 6' }) as string[]).join('');
+        displayNames[client.id] = { userid: client.id, displayName, color: generateRandomColor() };
+        client.emit(Constants.DISPLAY_NAME, displayNames[client.id]);
+        io.emit(Constants.USERS, displayNames);
+    });
+
+    /**
+     * Received when client presses 'Create Room' button with relevant information.
+     */
+    client.on(Constants.CREATE_ROOM, (data: Types.CreateRoom) => {
+        let roomid = uuid.v1(); // Generate random time-based id
+        // In case id already exists (highly unlikely in theory)
+        while (roomid in rooms) {
+            roomid = uuid.v1();
+        }
+        rooms[roomid] = {
+            owner: client.id,
+            size: data.size,
+            connected: {},
+        };
+
+        client.emit(Constants.CREATE_ROOM_SUCCESS, { roomid });
+        logInfo(`${client.id} has created room with id ${roomid}`);
+    });
+
+    /**
+     * Called when user goes directly to room link or right after owner has created room.
+     */
+    client.on(Constants.CONNECT_TO_ROOM, (data: Types.ConnectRoom) => {
+        /**
+         * Right now, since userid = socketid, a user can still technically open multiple rooms,
+         * each with a different socketid (multiple tabs).
+         * But a socketid should only be associated with one room.
+         */
+        if (userRooms[client.id]) {
+            logError(
+                `${displayNames[client.id]} is already connected to room ${
+                    userRooms[client.id]
+                } but tried to access room ${data.roomid}`,
+            );
+            return;
+        }
+
+        // FIXME: I don't know the better way to do this if statement in .ts (if obj, get obj.props)
+        if (rooms[data.roomid]) {
+            const room = rooms[data.roomid];
+            if (Object.keys(room.connected).length < room.size) {
+                room.connected[client.id] = displayNames[client.id]; // Update user in rooms object
+                userRooms[client.id] = data.roomid; // Also update user in users object (redundant)
+                logInfo(`${displayNames[client.id]} connected to room ${data.roomid}`);
+                Object.keys(room.connected).forEach(socketid => {
```

Comment:
You do `Object.keys(room.connected)` twice. Just make it a variable before the first time you do it.

Reply:
jk three times